src/txn/txn_ckpt.c

```diff
@@ -828,6 +828,9 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
 
     __wt_writeunlock(session, &txn_global->rwlock);
 
+    /* Wait for the commit generation to drain before bumping the snapshot. */
```

Comment:
I'm with Chenhao here - I'm concerned (mostly due to recent issues with things being blocked on the schema lock in the field) that doing this while holding the schema lock will be disruptive for some user(s). Especially if we can't avoid it for prepared transactions.

I think this needs to happen after the transaction ID has been allocated for the checkpoint, but I don't see why it needs to happen while holding the schema lock (except for how the code is currently structured.

I went digging into why the schema lock is held while setting up transaction context, and it was neither accidental nor definitely required:
https://github.com/wiredtiger/wiredtiger/pull/3319#discussion_r104345436

Reply:
I think if we wait for the commit generation to drain after we bump the snapshot then it is not correct because we want to ensure that all the things committed with that timestamp have really been committed and visible to us before we bump the snapshot. 

I think I understand your concern here that waiting for the generation to drain after holding the schema lock will enlarge the time of holding the schema lock. 

One possibility is to break the function `__checkpoint_prepare` into two stages to use the schema lock only when it is needed. I believe we need the schema lock after this [line](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn_ckpt.c#L851).
1. Call `__checkpoint_prepare_1` without schema lock till [this](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn_ckpt.c#L851) line
2. Call a new function `__checkpoint_prepare_2` with schema lock after [this](https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn_ckpt.c#L851) line.

PS - The function names used are just for reference.

@agorrod Please let me know your thoughts.

Reply:
Add to @raviprakashgiri29 's comment, we need to wait for the generation to gen before checkpoint bump the snapshot, otherwise, the following scenario is possible:

1. We start to commit, enter the generation and set the commit timestamp to 20.
2. We advance the stable timestamp to 30.
3. Checkpoint starts and get the snapshot. The transaction that is committing is not visible to the checkpoint's snapshot.
4. The transaction successfully commits.
5. Checkpoint waits the generation to gen but the checkpoint will not include that transaction as it is not visible to its snapshot.

As shown before, we must wait the generation to drain before we bump checkpoint's snapshot.